RawSpeed/DngDecoderSlices.cpp

```diff
@@ -143,6 +143,204 @@ my_error_throw (j_common_ptr cinfo)
+static void decodeFPDeltaRow(unsigned char * src, unsigned char * dst, size_t tileWidth, size_t realTileWidth, unsigned int bytesps, int factor)
+{
+  // DecodeDeltaBytes
+  for (size_t col = factor; col < realTileWidth*bytesps; ++col) {
+    src[col] += src[col - factor];
+  }
+  // Reorder bytes into the image
+  // 16 and 32-bit versions depend on local architecture, 24-bit does not
+  if (bytesps == 3) {
+    for (size_t col = 0; col < tileWidth; ++col) {
+      dst[col*3] = src[col];
+      dst[col*3 + 1] = src[col + realTileWidth];
+      dst[col*3 + 2] = src[col + realTileWidth*2];
+    }
+  } else {
+    union X { unsigned int x; unsigned char c; };
+    if (((union X){1}).c) {
+    for (size_t col = 0; col < tileWidth; ++col) {
+      for (size_t byte = 0; byte < bytesps; ++byte)
+        dst[col*bytesps + byte] = src[col + realTileWidth*(bytesps-byte-1)];  // Little endian
+    }
+    } else {
+    for (size_t col = 0; col < tileWidth; ++col) {
+      for (size_t byte = 0; byte < bytesps; ++byte)
+        dst[col*bytesps + byte] = src[col + realTileWidth*byte];
+        }
+      }
+    }
+}
+
+static uint32_t fp16ToFloat(uint16_t fp16)
+{
+  // IEEE-754-2008: binary16:
+  // bit 15 - sign
+  // bits 14-10 - exponent (5 bit)
+  // bits 9-0 - fraction (10 bit)
+  //
+  // exp = 0, fract = +-0: zero
+  // exp = 0; fract != 0: subnormal numbers
+  //                      equation: -1 ^ sign * 2 ^ -14 * 0.fraction
+  // exp = 1..30: normalized value
+  //              equation: -1 ^ sign * 2 ^ (exponent - 15) * 1.fraction
+  // exp = 31, fract = +-0: +-infinity
+  // exp = 31, fract != 0: NaN
+
+  uint32_t sign = (fp16 >> 15) & 1;
+  uint32_t fp16_exponent = (fp16 >> 10) & ((1 << 5) - 1);
+  uint32_t fp16_fraction = fp16 & ((1 << 10) - 1);
+
+  // Normalized or zero
+  // binary32 equation: -1 ^ sign * 2 ^ (exponent - 127) * 1.fraction
+  // => exponent32 - 127 = exponent16 - 15, exponent32 = exponent16 + 127 - 15
+  uint32_t fp32_exponent = fp16_exponent + 127 - 15;
+  uint32_t fp32_fraction = fp16_fraction << (23 - 10); // 23 is binary32 fraction size
+
+  if (fp16_exponent == 31) {
+    // Infinity or NaN
+    fp32_exponent = 255;
+  } else if (fp16_exponent == 0 && fp16_fraction == 0) {
+    // +-Zero
+    fp32_exponent = 0;
+    fp32_fraction = 0;
+  } else if (fp16_exponent == 0 && fp16_fraction != 0) {
+    // Subnormal numbers
+    // binary32 equation: -1 ^ sign * 2 ^ (exponent - 127) * 1.fraction
+    // binary16 equation: -1 ^ sign * 2 ^ -14 * 0.fraction, we can represent it
+    // as a normalized value in binary32, we have to shift fraction until we get 1.new_fraction
+    // and decrement exponent for each shift
+    fp32_exponent = -14 + 127;
+    while (!(fp32_fraction & (1 << 23))) {
+      fp32_exponent -= 1;
+      fp32_fraction <<= 1;
+    }
+    fp32_fraction &= ((1 << 23) - 1);
+  }
+  return (sign << 31) | (fp32_exponent << 23) | fp32_fraction;
+}
+
+static uint32_t fp24ToFloat(uint32_t fp24)
```

Comment:
Question: have you tried if hinting your compiler to `inline` this function helps him to actually to it? Removing a function call in the innermost loop might be worth it.